CLDC-3857 Add new group and profit status questions to organisation setup

Dinssa opened the pull request:
This PR adds new questions to an organisation, allowing support users to answer:
1) If the organisation is part of a group. And if it is, to set which group it is part of.
2) If the organisation is For-profit, Non-profit or neither because its a Local authority.

<img width="600" alt="Screenshot 2025-03-25 at 14 18 05" src="https://github.com/user-attachments/assets/ca0629ae-969f-4204-ad3e-b465d0f7a2fe" />

There is some JS functionality added to improve the user experience by preselecting options linked to other questions, this is optional and the page remains accessible should JS not be available. Validations have been put in place to prevent invalid options being selected as a fallback.

Dinssa committed:
Add questions to page, wip

Dinssa committed:
Add select as conditional question in radio like in logs

Dinssa committed:
Add existing value

Dinssa committed:
Add blank default

Dinssa committed:
Add values to about this organisation page

Dinssa committed:
Move grabbing values out of view

Dinssa committed:
Update new org page with js

Dinssa committed:
Selecting the profit status dependent on the provider type

Dinssa committed:
Add to org edit page for support users

Dinssa committed:
Add group logic

Dinssa committed:
Fix bug on existing orgs when group remains unchanged

Dinssa committed:
Clear group details if no longer part of a group

Dinssa committed:
Merge branch 'main' into CLDC-3857-Add-new-questions-to-organisation-setup

Dinssa committed:
Lint

Dinssa committed:
Merge remote-tracking branch 'origin/CLDC-3857-Add-new-questions-to-organisation-setup' into CLDC-3857-Add-new-questions-to-organisation-setup

Dinssa committed:
JS Lint

Dinssa committed:
Restore schema

Dinssa committed:
Merge branch 'main' into CLDC-3857-Add-new-questions-to-organisation-setup

# Conflicts:
#	app/frontend/controllers/index.js

Dinssa committed:
Merge branch 'main' into CLDC-3857-Add-new-questions-to-organisation-setup

Dinssa committed:
Remove manual_address_selected

Dinssa committed:
Merge branch 'main' into CLDC-3857-Add-new-questions-to-organisation-setup

Dinssa committed:
Revert "Remove manual_address_selected"

This reverts commit 812787fd09e45433508dfeda276ff57b1410f74d.

Dinssa committed:
Merge remote-tracking branch 'origin/CLDC-3857-Add-new-questions-to-organisation-setup' into CLDC-3857-Add-new-questions-to-organisation-setup

Dinssa committed:
Revert "Revert "Remove manual_address_selected""

This reverts commit 8ebd56036fd87a89844e8657e39d4d23742fd6d8.

Dinssa committed:
Update migration

Dinssa committed:
Merge branch 'main' into CLDC-3857-Add-new-questions-to-organisation-setup

Dinssa committed:
Remove file

Dinssa committed:
Merge remote-tracking branch 'origin/CLDC-3857-Add-new-questions-to-organisation-setup' into CLDC-3857-Add-new-questions-to-organisation-setup

Dinssa committed:
Merge branch 'main' into CLDC-3857-Add-new-questions-to-organisation-setup

# Conflicts:
#	db/schema.rb

Dinssa committed:
Merge branch 'main' into CLDC-3857-Add-new-questions-to-organisation-setup

Dinssa committed:
Merge branch 'main' into CLDC-3857-Add-new-questions-to-organisation-setup

Dinssa committed:
Move update after save

Dinssa committed:
Move methods to controller

Dinssa committed:
Update xml export

Dinssa committed:
Lint

Dinssa committed:
Update test xml file

Dinssa committed:
Update current user logic

Dinssa committed:
Lint

Dinssa committed:
Lint

Dinssa committed:
Reorder org js controller

Dinssa committed:
Lint

Dinssa committed:
Add tests to helper

Dinssa committed:
Some tests on the organisation pages

Dinssa commented on app/services/exports/organisation_export_service.rb:
```diff
@@ -63,8 +63,7 @@ def apply_cds_transformation(organisation)
       attribute_hash["merge_date"] = organisation.merge_date&.iso8601
       attribute_hash["available_from"] = organisation.available_from&.iso8601
-      attribute_hash["profit_status"] = nil # will need update when we add the field to the org
-      attribute_hash["group"] = nil # will need update when we add the field to the org
+      attribute_hash["profit_status"] = organisation.profit_status_before_type_cast
       attribute_hash["status"] = organisation.status
```
`group` should be passed in as one of the attributes on the organisation & I reassign `profit_status` just like with `provider_type`.

Dinssa commented on app/views/organisations/new.html.erb:
```diff
@@ -48,6 +48,8 @@
         :id,
         :name,
         label: { text: "Organisation type", size: "m" },
+        "data-controller": "organisations",
+        "data-action": "change->organisations#updateProfitStatusOptions",
```
Depending on the user's selection for the provider type question we will filter out invalid options from the profit status question using js. If js is disabled there are validations on the organisation model to prevent adding invalid options.

Dinssa commented on app/views/organisations/new.html.erb:
```diff
@@ -56,6 +58,31 @@
+          :name,
+          label: { text: "Is the organisation for-profit?", size: "m" },
+          options: { disabled: [""], selected: @organisation.profit_status || "" } %>
+
```
This adds a Yes/No question of whether the organisation is part of a group or not. If Yes is selected a additional question for find and select another organisation that is part of the same group is asked. In the options organisations with groups will have the group appended to their name e.g. "MHCLG (Group2)", those without will just have their name e.g. "MHCLG". User's just need to select any organisation already in the group, should there be multiple.

Dinssa commented on app/views/organisations/edit.html.erb:
```diff
@@ -33,6 +33,33 @@
+            <%= f.govuk_collection_select :group_member_id,
+                group_organisation_options,
+                :id,
+                :name,
+                label: { text: "Search for an organisation that is part of the same group as this organisation", size: "m" },
+                options: { disabled: [""], selected: @organisation.oldest_group_member&.id || "" },
+                "data-controller": %w[accessible-autocomplete conditional-filter] %>
+          <% end %>
```
In the org edit page, should a org be part of a group we'll show the search box to find an organisation. I chose to make the organisation shown in the box the oldest organisation within the group, rather than a random organisation or the org a user may have selected when creating the org. The idea being that the oldest would be sort of the "leader" and to keep things the same when viewing the editable question across multiple orgs in a group. This is only visible to support users who are the only users that can change a group.

kosiakkatrina replied:
It's kinda weird tbh that we're selecting a single org, but might be joining a group with many

kosiakkatrina replied:
If there's lots of orgs in 1 group is there any way to know what all of the orgs in that group are?

Dinssa replied:
The search result will return the organisation name with the group in brackets at the end, if present. You can search by the text within the brackets. For example, if the organisation MHCLG is in group 2, the search box will store it as "MHCLG (GROUP2)." You can then search for either "MHCLG" or "GROUP2" - to retrieve all organisations that are also in group 2 in the dropdown list.

Dinssa commented on app/controllers/organisations_controller.rb:
```diff
@@ -385,4 +389,19 @@ def get_duplicate_schemes_and_locations
     end
   end
+
+  def assign_group_number(current_org_id, selected_org_id)
+    selected_org = Organisation.find_by(id: selected_org_id)
+    if selected_org&.group.present?
+      selected_org.group
+    else
+      next_group_number = next_available_group_number
+      selected_org.update!(group_member: true, group_member_id: current_org_id, group: next_group_number) if selected_org
+      next_group_number
+    end
+  end
```
When creating or editing an org we select another org that the org should be in the same group with. If the selected org has a group number we'll take that. If not, we'll give them both the next available group number.

kosiakkatrina commented on app/controllers/organisations_controller.rb:
```diff
@@ -91,6 +91,7 @@ def create
     selected_rent_periods = rent_period_params[:rent_periods].compact_blank
     @organisation = Organisation.new(org_params)
     if @organisation.save
+      @organisation.update!(group: assign_group_number(@organisation.id, org_params[:group_member_id])) if org_params[:group_member]
```
Could we just add the group to the org_params? Something similar to what we have in locations_controller > location_params ?
That way we'd avoid saving the org twice

Dinssa replied:
This approach allows us to use the `assign_group_number` method to find the next available group number and simultaneously add the group number to the other selected organisation. This is primarily to ensure the group number is added to the other selected organisation after a user has confirmed that this is the group they want to select .ie. after saving. Performing this action in the view would be premature before the user has chosen to create the organisation.

kosiakkatrina commented on app/controllers/organisations_controller.rb:
```diff
@@ -127,6 +128,9 @@ def reactivate
     if (current_user.data_coordinator? && org_params[:active].nil?) || current_user.support?
+      if org_params[:group_member] && org_params[:group_member_id]
+        @organisation.group = assign_group_number(@organisation.id, org_params[:group_member_id])
+      end
```
Same here, this could probably be removed if we add group to org_params

kosiakkatrina replied:
What if we select no? Should it remove the group number from the other org?

Dinssa replied:
If you select "No," the other organisation selection question should be empty, as it will be disabled, and the other organisation should not be affected. On the edit page, if you change an organisation to not be part of a group, there is logic to clear the group value, but only for that specific log.

kosiakkatrina replied:
Sorry, I mean what if you select yes, choose an org and then select no afterwards?

Dinssa replied:
It should not remove the group number from the other org, if someone did select no after choosing an org, we should not assign any group number to the new organisation and I don't think we will as it's currently implemented. Please check @samyou-softwire

samyou-softwire replied:
yup the `if org_params[:group_member]` on that line (and on the save line) should prevent an association being created

kosiakkatrina commented on app/controllers/organisations_controller.rb:
```diff
@@ -385,4 +389,19 @@ def get_duplicate_schemes_and_locations
+      selected_org.group
+    else
+      next_group_number = next_available_group_number
+      selected_org.update!(group_member: true, group_member_id: current_org_id, group: next_group_number) if selected_org
```
What would happen if we fail to save the current org because of a different validation? Would that still set the group number on the selected org?

Dinssa replied:
@samyou-softwire I didn't get a chance to have a look into this, could you please investigate.

samyou-softwire replied:
found that:
- on create `assign_group_number` is guarded by a `.save` call so will not associate if validation fails
- on update `assign_group_number` was outside the update check so would associate if a validation fails. this has been resolved

Dinssa committed:
Merge branch 'main' into CLDC-3857-Add-new-questions-to-organisation-setup

Dinssa committed:
Fix label text for profit status question in forms

Dinssa committed:
Add CSV task to update organisation profit status and group member fields

Dinssa commented on lib/tasks/update_organisations_group_profit_status.rake:
```diff
@@ -0,0 +1,29 @@
+          profit_status: row["profit_status"].to_i,
+          group_member: true,
+          group: row["group"].to_i,
+        )
```
@samyou-softwire Had a look at the CSV file shared and I released this task does not account for updating organisations without a group, where we still want to update the profit status. I think this needs to be amended to conditionally update the `group` and `group_member` fields only if a group number is provided, for example we do not want to be setting `group_member` to `true` for each of the logs in the file.

Additionally, I would remove the `core_registration_number` and `core_type` from the CSV before adding it in to the branch. The `core_registration_number` is not needed, nor is `core_type`. However, there is a validation that checks the profit status against the core_type (e.g., profit status must be local authority if the `core_type` is LA). This validation does not require the column to be included in the CSV, it should already be set.

samyou-softwire replied:
have imported the csv accordingly

samyou-softwire committed:
only associate group on update if validation passes

samyou-softwire committed:
Update import raketask to use provided csv

looking into it I couldn't see a use for skip_group_member_validation. it temporarily disables validation but since it's not saved to database it'll make the organisation invalid in the future

opted instead to make all properties valid when importing

samyou-softwire committed:
Update tests to check more cases of organisation import

samyou-softwire committed:
Make XML tests check unordered file contents

stops tests flakily failing on the order of cols which is not important in xml
